url8f84d8d956_klapaukh_JTuttePoly_tgz-pJ8-BitTutteJ8/src/util/Isomorphism.java

```diff
@@ -0,0 +1,389 @@
+			System.out.println(partition);
+			System.out.println(resultTriple);
+			throw e;
+		}
+		// System.out.println();
+		// print(resultTriple.first, resultTriple.second);
+		return resultTriple.third;
+	}
+
+	/**
+	 * Triple<Partition, value, edges>
+	 * 
+	 * @param g
+	 * @param partition
+	 * @param soFar
+	 * @return
+	 */
+	private static Triple<List<Set<Integer>>, int[], int[]> searchTree(Graph g, List<Set<Integer>> partition,
+			Triple<List<Set<Integer>>, int[], int[]> soFar) {
+		if (discrete(partition)) {
+			int[] label = partitionToMatrix(partition, g.domainSize(), g);
+			// int[] label = partitionToLabel(partition, g.domainSize());
+			if (soFar == null) {
+				return new Triple<List<Set<Integer>>, int[], int[]>(partition, label, label);
+			} else {
+				// int[] value = partitionToValue(partition, g, label);
+				if (greaterThan(label, soFar.second)) {
+					return new Triple<List<Set<Integer>>, int[], int[]>(partition, label, label);
+				}
+				return soFar;
+			}
+		}
+		// Find the first smallest non trivial cell of partition since there
+		// must be one
+		int min = -1;
+		int size = Integer.MAX_VALUE;
+		for (int i = 0; i < partition.size(); i++) {
+			int s = partition.get(i).size();
+			if (s > 1 && s < size) {
+				min = i;
+				size = s;
+			}
+		}
+		Set<Integer> Wk = partition.get(min);
+
+		Triple<List<Set<Integer>>, int[], int[]> current = soFar;
+		for (int i : Wk) {
+			// DO THE splitting
+			List<Set<Integer>> newPartition = new ArrayList<Set<Integer>>();
+
+			for (int j = 0; j < min; j++) {
+				newPartition.add(partition.get(j));
+			}
+
+			Set<Integer> u = new TreeSet<Integer>();
+			u.add(i);
+			newPartition.add(u);
+			Set<Integer> Wk1 = new TreeSet<Integer>(Wk);
+			Wk1.remove(i);
+			newPartition.add(Wk1);
+
+			for (int j = min + 1; j < partition.size(); j++) {
+				newPartition.add(partition.get(j));
+			}
+
+			// Now you need to do the equitable refinement again
+			List<Set<Integer>> alpha = new ArrayList<Set<Integer>>();
+			alpha.add(u);
+			newPartition = equitableRefinement(g, newPartition, alpha);
+
+			// if (discrete(newPartition)) {
+			// int[] value = partitionToMatrix(newPartition, g.domainSize(), g);
+			// print(newPartition, value);
+			// } else {
+			// print(newPartition, null);
+			// }
+
+			current = searchTree(g, newPartition, current);
+
+		}
+		return current;
+
+	}
+
+	private static boolean greaterThan(int[] value, int[] second) {
+		for (int i = 0; i < value.length; i++) {
+			if (greaterthan(0xFFFFFFFFL & value[i], 0xFFFFFFFFL & second[i])) {
+				return true;
+			} else if (lessthan(0xFFFFFFFFL & value[i], 0xFFFFFFFFL & second[i])) {
+				return false;
+			}
+		}
+		return false;
+	}
+
+	private static boolean greaterthan(long i, long j) {
+		return i > j;
+	}
+
+	private static boolean lessthan(long i, long j) {
+		return i < j;
+	}
+
+	private static List<Set<Integer>> equitableRefinement(Graph g, List<Set<Integer>> partition, List<Set<Integer>> alpha) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.